### PR TITLE
refactor(types): using options pattern for memo parameter on new tx functions

### LINKED ISSUE #1440: Using Option Pattern to Make `memo` Optional in Creating New Transactions
## Describe the problem you'd like to have solved

This issue proposes a modification to the [newTx](https://github.com/pactus-project/pactus/blob/b850baa9afc0ac114fccc03229bcf8e27c8b1a7c/types/tx/tx.go#L45) function to use the option pattern for the `memo` parameter.
Currently, the `memo` parameter  is required, which might not always be necessary for every transaction. Using the option pattern will provide more flexibility and make the code cleaner.

This may clean up tests, since many `memo` fields are duplicated. It also helps to add more options in the future, such as `fee`, and others.

### Describe the ideal solution

The proposal is to use the option pattern to make `memo` an optional parameter. Here is a potential implementation:

```go
type TxOption func(*txData)

func WithMemo(memo string) TxOption {
    return func(td *txData) {
        td.Memo = memo
    }
}

func newTx(lockTime uint32, pld payload.Payload, fee amount.Amount, opts ...TxOption) *Tx {
    data := txData{
        Flags:    flagNotSigned,
        LockTime: lockTime,
        Version:  versionLatest,
        Payload:  pld,
        Fee:      fee,
    }

    for _, opt := range opts {
        opt(&data)
    }

    return &Tx{data: data}
}
```

### PR DESCRIPTION
**Pull request by maxipaz**
## Description

Making `memo` field optional in the `newTx` and its callers functions, by using options pattern.

## Related issue(s)

- Fixes #1440

### CODE REVIEW

File: execution/execution_test.go [hunk 1]
```diff
@@ -173,7 +173,7 @@ func TestSubsidyLockTime(t *testing.T) {
 
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
-			trx := tx.NewSubsidyTx(tc.lockTime, ts.RandAccAddress(), 1000, "subsidy-test")
+			trx := tx.NewSubsidyTx(tc.lockTime, ts.RandAccAddress(), 1000, tx.WithMemo("subsidy-test"))
```
Comment (b00f) [inline execution/execution_test.go:176]:
```suggestion
			trx := tx.NewSubsidyTx(tc.lockTime, ts.RandAccAddress(), 1000)
```
One of the advantages of making "memo" optional is to prevent some duplicated declarations in tests. This helps keep the tests clean. 

An example here:
The `TestSubsidyLockTime` is checking the "Lock Time for the Subsidy Transactions." The test name itself is clear, so we can remove the memo and keep the test simpler.

File: execution/execution_test.go [hunk 2]
```diff
@@ -236,14 +236,14 @@ func TestExecute(t *testing.T) {
 	t.Run("Invalid transaction, Should return error", func(t *testing.T) {
 		randAddr := ts.RandAccAddress()
 		trx := tx.NewTransferTx(lockTime, randAddr, ts.RandAccAddress(),
-			ts.RandAmount(), ts.RandFee(), "invalid-tx")
+			ts.RandAmount(), ts.RandFee(), tx.WithMemo("invalid-tx"))
```
Comment (b00f) [inline execution/execution_test.go:239]:
```suggestion
			ts.RandAmount(), ts.RandFee())
```


Same here. The test declares that this is an `invalid-tx`.

I don't mention other similar cases.

File: store/store_test.go [hunk 3]
```diff
@@ -75,19 +75,19 @@ func TestBlockHash(t *testing.T) {
 
 	sb, _ := td.store.Block(1)
 
-	assert.Equal(t, hash.UndefHash, td.store.BlockHash(0))
-	assert.Equal(t, hash.UndefHash, td.store.BlockHash(util.MaxUint32))
-	assert.Equal(t, sb.BlockHash, td.store.BlockHash(1))
+	assert.Equal(t, td.store.BlockHash(0), hash.UndefHash)
```
Comment (b00f) [inline store/store_test.go:78]: Expected value should come first. So please revert this change.

File: store/store_test.go [hunk 4]
```diff
@@ -75,19 +75,19 @@ func TestBlockHash(t *testing.T) {
...
 
 func TestBlockHeight(t *testing.T) {
 	td := setup(t, nil)
 
 	sb, _ := td.store.Block(1)
 
-	assert.Equal(t, uint32(0), td.store.BlockHeight(hash.UndefHash))
-	assert.Equal(t, uint32(0), td.store.BlockHeight(td.RandHash()))
-	assert.Equal(t, uint32(1), td.store.BlockHeight(sb.BlockHash))
+	assert.Equal(t, td.store.BlockHeight(hash.UndefHash), uint32(0))
```
Comment (b00f) [inline store/store_test.go:88]: Expected value should come first. So please revert this change.

File: store/store_test.go [hunk 5]
```diff
@@ -140,9 +140,9 @@ func TestIndexingPublicKeys(t *testing.T) {
 			assert.NoError(t, err)
 
 			if addr.IsAccountAddress() {
-				assert.Equal(t, addr, pubKey.AccountAddress())
```
Comment (b00f) [inline store/store_test.go:143]: I think your base branch is out-dated. These changes should be reverted.

File: txpool/txpool_test.go [hunk 6]
```diff
@@ -70,7 +70,7 @@ func (td *testData) shouldPublishTransaction(t *testing.T, id tx.ID) {
 
 			if msg.Type() == message.TypeTransaction {
 				m := msg.(*message.TransactionsMessage)
-				assert.Equal(t, id, m.Transactions[0].ID())
+				assert.Equal(t, m.Transactions[0].ID(), id)
```
Comment (b00f) [inline txpool/txpool_test.go:73]: same here

**Review by b00f — CHANGES_REQUESTED** [review]

File: store/store_test.go (hunk 3 above)
Comment (maxipaz) [reply store/store_test.go]: Yes, my branch was outdated. Already fixed. Thanks for the warning 🙂

File: execution/execution_test.go (hunk 1 above)
Comment (maxipaz) [reply execution/execution_test.go]: Good suggestion, will make them simpler, yes.

File: execution/executor/bond_test.go [hunk 7]
```diff
@@ -159,23 +150,23 @@ func TestSmallBond(t *testing.T) {
 
 	t.Run("Rejects bond transaction with zero amount", func(t *testing.T) {
 		trx := tx.NewBondTx(lockTime, senderAddr,
-			receiverAddr, nil, 0, fee, "attacking validator")
+			receiverAddr, nil, 0, fee, tx.WithMemo("attacking validator"))
```
Comment (b00f) [inline execution/executor/bond_test.go:153]:
```suggestion
			receiverAddr, nil, 0, fee)
```

File: execution/executor/bond_test.go [hunk 8]
```diff
@@ -159,23 +150,23 @@ func TestSmallBond(t *testing.T) {
...
 
 		td.check(t, trx, true, SmallStakeError{td.sandbox.TestParams.MinimumStake})
 		td.check(t, trx, false, SmallStakeError{td.sandbox.TestParams.MinimumStake})
 	})
 
 	t.Run("Rejects bond transaction below full validator stake", func(t *testing.T) {
 		trx := tx.NewBondTx(lockTime, senderAddr,
-			receiverAddr, nil, 1, fee, "attacking validator")
+			receiverAddr, nil, 1, fee, tx.WithMemo("attacking validator"))
```
Comment (b00f) [inline execution/executor/bond_test.go:161]:
```suggestion
			receiverAddr, nil, 1, fee)
```

File: execution/executor/bond_test.go [hunk 9]
```diff
@@ -159,23 +150,23 @@ func TestSmallBond(t *testing.T) {
...
 
 		td.check(t, trx, true, SmallStakeError{td.sandbox.TestParams.MinimumStake})
 		td.check(t, trx, false, SmallStakeError{td.sandbox.TestParams.MinimumStake})
 	})
 
 	t.Run("Accepts bond transaction reaching full validator stake", func(t *testing.T) {
 		trx := tx.NewBondTx(lockTime, senderAddr,
-			receiverAddr, nil, 2, fee, "fulfilling validator stake")
+			receiverAddr, nil, 2, fee, tx.WithMemo("fulfilling validator stake"))
```
Comment (b00f) [inline execution/executor/bond_test.go:169]:
```suggestion
			receiverAddr, nil, 2, fee)
```

File: execution/executor/bond_test.go [hunk 10]
```diff
@@ -184,7 +175,7 @@ func TestSmallBond(t *testing.T) {
 
 	t.Run("Rejects bond transaction with zero amount on full validator", func(t *testing.T) {
 		trx := tx.NewBondTx(lockTime, senderAddr,
-			receiverAddr, nil, 0, fee, "attacking validator")
+			receiverAddr, nil, 0, fee, tx.WithMemo("attacking validator"))
```
Comment (b00f) [inline execution/executor/bond_test.go:178]:
```suggestion
			receiverAddr, nil, 0, fee)
```

File: execution/executor/unbond_test.go [hunk 11]
```diff
@@ -46,22 +46,22 @@ func TestExecuteUnbondTx(t *testing.T) {
...
+		trx := tx.NewUnbondTx(lockTime, valAddr)
 
 		td.check(t, trx, true, nil)
 		td.check(t, trx, false, nil)
 		td.execute(t, trx)
 	})
 
 	t.Run("Should fail, Cannot unbond if already unbonded", func(t *testing.T) {
-		trx := tx.NewUnbondTx(lockTime, valAddr, "Ok")
+		trx := tx.NewUnbondTx(lockTime, valAddr, tx.WithMemo("Ok"))
```
Comment (b00f) [inline execution/executor/unbond_test.go:64]:
```suggestion
		trx := tx.NewUnbondTx(lockTime, valAddr)
```

File: execution/executor/unbond_test.go [hunk 12]
```diff
@@ -87,7 +87,7 @@ func TestPowerDeltaUnbond(t *testing.T) {
 	val.AddToStake(amt)
 	td.sandbox.UpdateValidator(val)
 	lockTime := td.sandbox.CurrentHeight()
-	trx := tx.NewUnbondTx(lockTime, valAddr, "Ok")
+	trx := tx.NewUnbondTx(lockTime, valAddr, tx.WithMemo("Ok"))
```
Comment (b00f) [inline execution/executor/unbond_test.go:90]:
```suggestion
	trx := tx.NewUnbondTx(lockTime, valAddr)
```

File: execution/executor/withdraw_test.go [hunk 13]
```diff
@@ -32,16 +32,15 @@ func TestExecuteWithdrawTx(t *testing.T) {
...
+		trx := tx.NewWithdrawTx(lockTime, randomAddr, receiverAddr, amt, fee)
 
 		td.check(t, trx, true, ValidatorNotFoundError{Address: randomAddr})
 		td.check(t, trx, false, ValidatorNotFoundError{Address: randomAddr})
 	})
 
 	t.Run("Should fail, hasn't unbonded yet", func(t *testing.T) {
 		trx := tx.NewWithdrawTx(lockTime, senderAddr, receiverAddr,
-			amt, fee, "need to unbond first")
+			amt, fee, tx.WithMemo("need to unbond first"))
```
Comment (b00f) [inline execution/executor/withdraw_test.go:43]:
```suggestion
			amt, fee)
```

File: execution/executor/withdraw_test.go [hunk 14]
```diff
@@ -71,7 +68,7 @@ func TestExecuteWithdrawTx(t *testing.T) {
 
 	t.Run("Should pass, Everything is Ok!", func(t *testing.T) {
 		trx := tx.NewWithdrawTx(lockTime, senderAddr, receiverAddr,
-			amt, fee, "should be able to withdraw the stake")
+			amt, fee, tx.WithMemo("should be able to withdraw the stake"))
```
Comment (b00f) [inline execution/executor/withdraw_test.go:71]:
```suggestion
			amt, fee)
```

File: txpool/txpool_test.go [hunk 15]
```diff
@@ -167,16 +166,16 @@ func TestPrepareBlockTransactions(t *testing.T) {
 	td.sandbox.UpdateValidator(val3)
 
 	transferTx := tx.NewTransferTx(randHeight+1, acc1Addr,
-		td.RandAccAddress(), 1e9, 100_000_000, "transfer-tx")
+		td.RandAccAddress(), 1e9, 100_000_000, tx.WithMemo("transfer-tx"))
```
Comment (b00f) [inline txpool/txpool_test.go:169]:
```suggestion
		td.RandAccAddress(), 1e9, 100_000_000)
```

File: txpool/txpool_test.go [hunk 16]
```diff
@@ -167,16 +166,16 @@ func TestPrepareBlockTransactions(t *testing.T) {
...
 
 	pub, _ := td.RandBLSKeyPair()
 	bondTx := tx.NewBondTx(randHeight+2, acc1Addr,
-		pub.ValidatorAddress(), pub, 1e9, 100_000_000, "bond-tx")
+		pub.ValidatorAddress(), pub, 1e9, 100_000_000, tx.WithMemo("bond-tx"))
```
Comment (b00f) [inline txpool/txpool_test.go:173]:
```suggestion
		pub.ValidatorAddress(), pub, 1e9, 100_000_000)
```

File: txpool/txpool_test.go [hunk 17]
```diff
@@ -167,16 +166,16 @@ func TestPrepareBlockTransactions(t *testing.T) {
...
 
-	unbondTx := tx.NewUnbondTx(randHeight+3, val1.Address(), "unbond-tx")
+	unbondTx := tx.NewUnbondTx(randHeight+3, val1.Address(), tx.WithMemo("unbond-tx"))
```
Comment (b00f) [inline txpool/txpool_test.go:175]:
```suggestion
	unbondTx := tx.NewUnbondTx(randHeight+3, val1.Address())
```

File: txpool/txpool_test.go [hunk 18]
```diff
@@ -167,16 +166,16 @@ func TestPrepareBlockTransactions(t *testing.T) {
...
 
 	withdrawTx := tx.NewWithdrawTx(randHeight+4, val2.Address(),
-		td.RandAccAddress(), 1e9, 100_000_000, "withdraw-tx")
+		td.RandAccAddress(), 1e9, 100_000_000, tx.WithMemo("withdraw-tx"))
```
Comment (b00f) [inline txpool/txpool_test.go:178]:
```suggestion
		td.RandAccAddress(), 1e9, 100_000_000)
```

File: txpool/txpool_test.go [hunk 19]
```diff
@@ -202,7 +201,7 @@ func TestAppendAndBroadcast(t *testing.T) {
 
 	height := td.RandHeight()
 	td.sandbox.TestStore.AddTestBlock(height)
-	testTrx := tx.NewSubsidyTx(height+1, td.RandAccAddress(), 1e9, "subsidy-tx")
+	testTrx := tx.NewSubsidyTx(height+1, td.RandAccAddress(), 1e9, tx.WithMemo("subsidy-tx"))
```
Comment (b00f) [inline txpool/txpool_test.go:204]:
```suggestion
	testTrx := tx.NewSubsidyTx(height+1, td.RandAccAddress(), 1e9)
```

File: types/tx/tx_test.go [hunk 20]
```diff
@@ -115,7 +115,7 @@ func TestBasicCheck(t *testing.T) {
 		invAddr := ts.RandAccAddress()
 		invAddr[0] = 3
 		trx := tx.NewTransferTx(ts.RandHeight(),
-			ts.RandAccAddress(), invAddr, 1e9, ts.RandAmount(), "invalid address")
+			ts.RandAccAddress(), invAddr, 1e9, ts.RandAmount(), tx.WithMemo("invalid address"))
```
Comment (b00f) [inline types/tx/tx_test.go:118]:
```suggestion
			ts.RandAccAddress(), invAddr, 1e9, ts.RandAmount())
```

File: types/tx/tx_test.go [hunk 21]
```diff
@@ -166,7 +162,7 @@ func TestBasicCheck(t *testing.T) {
 	t.Run("Invalid signer address", func(t *testing.T) {
 		valKey := ts.RandValKey()
 		trx := tx.NewTransferTx(ts.RandHeight(),
-			ts.RandAccAddress(), ts.RandAccAddress(), 1, 1, "invalid valKey")
+			ts.RandAccAddress(), ts.RandAccAddress(), 1, 1, tx.WithMemo("invalid valKey"))
```
Comment (b00f) [inline types/tx/tx_test.go:165]:
```suggestion
			ts.RandAccAddress(), ts.RandAccAddress(), 1, 1)
```

File: util/testsuite/testsuite.go [hunk 22]
```diff
@@ -530,7 +530,7 @@ func (ts *TestSuite) GenerateTestTransferTx(options ...func(tm *TransactionMaker
 		opt(tm)
 	}
 	trx := tx.NewTransferTx(tm.LockTime, tm.PubKey.AccountAddress(), ts.RandAccAddress(),
-		tm.Amount, tm.Fee, "test send-tx")
+		tm.Amount, tm.Fee, tx.WithMemo("test send-tx"))
```
Comment (b00f) [inline util/testsuite/testsuite.go:533]:
```suggestion
		tm.Amount, tm.Fee)
```

File: util/testsuite/testsuite.go [hunk 23]
```diff
@@ -544,7 +544,7 @@ func (ts *TestSuite) GenerateTestBondTx(options ...func(tm *TransactionMaker)) *
 		opt(tm)
 	}
 	trx := tx.NewBondTx(tm.LockTime, tm.PubKey.AccountAddress(), ts.RandValAddress(),
-		nil, tm.Amount, tm.Fee, "test bond-tx")
+		nil, tm.Amount, tm.Fee, tx.WithMemo("test bond-tx"))
```
Comment (b00f) [inline util/testsuite/testsuite.go:547]:
```suggestion
		nil, tm.Amount, tm.Fee)
```

File: util/testsuite/testsuite.go [hunk 24]
```diff
@@ -571,7 +571,7 @@ func (ts *TestSuite) GenerateTestUnbondTx(options ...func(tm *TransactionMaker))
 	for _, opt := range options {
 		opt(tm)
 	}
-	trx := tx.NewUnbondTx(tm.LockTime, tm.PubKey.ValidatorAddress(), "test unbond-tx")
+	trx := tx.NewUnbondTx(tm.LockTime, tm.PubKey.ValidatorAddress(), tx.WithMemo("test unbond-tx"))
```
Comment (b00f) [inline util/testsuite/testsuite.go:574]:
```suggestion
	trx := tx.NewUnbondTx(tm.LockTime, tm.PubKey.ValidatorAddress())
```

File: util/testsuite/testsuite.go [hunk 25]
```diff
@@ -585,7 +585,7 @@ func (ts *TestSuite) GenerateTestWithdrawTx(options ...func(tm *TransactionMaker
 		opt(tm)
 	}
 	trx := tx.NewWithdrawTx(tm.LockTime, tm.PubKey.ValidatorAddress(), ts.RandAccAddress(),
-		tm.Amount, tm.Fee, "test withdraw-tx")
+		tm.Amount, tm.Fee, tx.WithMemo("test withdraw-tx"))
```
Comment (b00f) [inline util/testsuite/testsuite.go:588]:
```suggestion
		tm.Amount, tm.Fee)
```

File: www/grpc/wallet_test.go [hunk 26]
```diff
@@ -161,7 +161,8 @@ func TestLoadWallet(t *testing.T) {
 
 	t.Run("Sign raw transaction, OK", func(t *testing.T) {
 		wltAddr, _ := crypto.AddressFromString(wltAddrInfo.Address)
-		bondTx := tx.NewBondTx(td.RandHeight(), wltAddr, td.RandValAddress(), nil, td.RandAmount(), td.RandAmount(), "memo")
+		bondTx := tx.NewBondTx(td.RandHeight(), wltAddr, td.RandValAddress(), nil, td.RandAmount(),
+			td.RandAmount(), tx.WithMemo("memo"))
```
Comment (b00f) [inline www/grpc/wallet_test.go:165]:
```suggestion
			td.RandAmount())
```

File: www/grpc/wallet_test.go [hunk 27]
```diff
@@ -183,7 +184,8 @@ func TestLoadWallet(t *testing.T) {
 
 	t.Run("Sign raw transaction using not loaded wallet", func(t *testing.T) {
 		wltAddr, _ := crypto.AddressFromString(wltAddrInfo.Address)
-		bondTx := tx.NewBondTx(td.RandHeight(), wltAddr, td.RandValAddress(), nil, td.RandAmount(), td.RandAmount(), "memo")
+		bondTx := tx.NewBondTx(td.RandHeight(), wltAddr, td.RandValAddress(), nil, td.RandAmount(),
+			td.RandAmount(), tx.WithMemo("memo"))
```
Comment (b00f) [inline www/grpc/wallet_test.go:188]:
```suggestion
			td.RandAmount())
```

**Review by b00f — CHANGES_REQUESTED** [review]

**Comment by b00f** [comment]
I proposed to remove some unnecessary memos from the test. Maybe you can align lines in some tests.

**Review by b00f — APPROVED** [review]
